{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/wiz124/chem169-git/blob/main/Li_Harry_RID_014_code.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "thvAeiTVVxWq",
        "outputId": "178e437a-e922-42e6-ed91-ffaa39cfb7e3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'MACHINELEARNING': {'residue_embeddings': array([[ 0.05606826, -0.02779197,  0.1984435 , ...,  0.29308623,\n",
            "        -0.0005211 , -0.001761  ],\n",
            "       [ 0.01368738,  0.08926699,  0.07153449, ...,  0.02676891,\n",
            "        -0.05958881,  0.08751322],\n",
            "       [ 0.0791374 ,  0.05739338,  0.05229965, ..., -0.12164824,\n",
            "        -0.2122045 , -0.03057704],\n",
            "       ...,\n",
            "       [-0.07602476,  0.17316677,  0.04309352, ..., -0.17556725,\n",
            "         0.18471679,  0.06163794],\n",
            "       [ 0.0032055 ,  0.07336932,  0.23836282, ..., -0.11572927,\n",
            "        -0.1422069 ,  0.17654972],\n",
            "       [ 0.1213042 ,  0.09664907,  0.15444879, ..., -0.22865772,\n",
            "        -0.07566334, -0.07008101]], dtype=float32), 'protein_embedding': array([ 2.7157145e-02,  6.7918792e-02,  1.1835212e-01, ...,\n",
            "       -3.1786315e-02,  6.4118205e-05,  1.0375285e-02], dtype=float32)}, 'MGCHINELEARNING': {'residue_embeddings': array([[-0.01221433, -0.04243858,  0.23617591, ...,  0.26975685,\n",
            "        -0.10064612,  0.03309401],\n",
            "       [ 0.1237796 , -0.01890814,  0.12185148, ...,  0.13858302,\n",
            "        -0.02139318,  0.07952821],\n",
            "       [ 0.0703975 , -0.04017996,  0.037202  , ..., -0.04747648,\n",
            "        -0.27260008,  0.03603963],\n",
            "       ...,\n",
            "       [-0.03181905,  0.18723276,  0.03050748, ..., -0.13183856,\n",
            "         0.15668514,  0.08637113],\n",
            "       [ 0.04160677,  0.09825741,  0.22365591, ..., -0.04876086,\n",
            "        -0.18204392,  0.16523133],\n",
            "       [ 0.11441068,  0.05866617,  0.15189052, ..., -0.16161145,\n",
            "        -0.11273676,  0.0340802 ]], dtype=float32), 'protein_embedding': array([ 0.03928996,  0.05962785,  0.12646566, ...,  0.01712848,\n",
            "       -0.03295295,  0.06646581], dtype=float32)}, 'MLCHINELEARNING': {'residue_embeddings': array([[-0.02348768, -0.06737775,  0.30448887, ...,  0.22063273,\n",
            "        -0.0216808 , -0.03832598],\n",
            "       [ 0.1255877 , -0.02136015,  0.20788512, ...,  0.07072802,\n",
            "         0.02024508, -0.0259651 ],\n",
            "       [ 0.13519073, -0.08404002,  0.0928096 , ..., -0.09306771,\n",
            "        -0.29415426,  0.03936787],\n",
            "       ...,\n",
            "       [-0.04450252,  0.18036602,  0.06166712, ..., -0.17303912,\n",
            "         0.16002263,  0.09177554],\n",
            "       [ 0.02496321,  0.1156003 ,  0.24694373, ..., -0.0935742 ,\n",
            "        -0.16940343,  0.20230731],\n",
            "       [ 0.08165168,  0.09079417,  0.12927775, ..., -0.1996663 ,\n",
            "        -0.06689113,  0.01048542]], dtype=float32), 'protein_embedding': array([ 0.04051566,  0.05903596,  0.15929994, ..., -0.01560776,\n",
            "       -0.02459794,  0.06557129], dtype=float32)}, 'MACHINEAEARNING': {'residue_embeddings': array([[ 0.04304859, -0.01717667,  0.20790304, ...,  0.28659803,\n",
            "        -0.02484175,  0.02266391],\n",
            "       [-0.01545749,  0.10129584,  0.05071112, ...,  0.03306917,\n",
            "        -0.08175289,  0.11969808],\n",
            "       [ 0.09800516,  0.04179817,  0.0534738 , ..., -0.13397202,\n",
            "        -0.22516078,  0.013848  ],\n",
            "       ...,\n",
            "       [-0.05305982,  0.2048952 ,  0.05135461, ..., -0.17196594,\n",
            "         0.19058347,  0.09969407],\n",
            "       [ 0.0415617 ,  0.12278973,  0.2219481 , ..., -0.09047426,\n",
            "        -0.15482979,  0.18269931],\n",
            "       [ 0.11829355,  0.08116539,  0.1686621 , ..., -0.22043261,\n",
            "        -0.10947494,  0.01592181]], dtype=float32), 'protein_embedding': array([ 0.03536478,  0.08991922,  0.11587524, ..., -0.02221408,\n",
            "       -0.01809251,  0.06115762], dtype=float32)}, 'MACHINEGEARNING': {'residue_embeddings': array([[-0.02047198, -0.05934278,  0.18795408, ...,  0.31268495,\n",
            "        -0.07668409,  0.04833537],\n",
            "       [-0.05122827,  0.09534466, -0.01075379, ...,  0.06221346,\n",
            "        -0.1297039 ,  0.14222705],\n",
            "       [ 0.0700347 ,  0.01101001,  0.0208775 , ..., -0.1051219 ,\n",
            "        -0.25953165,  0.08852682],\n",
            "       ...,\n",
            "       [-0.00587133,  0.173598  ,  0.04168514, ..., -0.07634631,\n",
            "         0.17283033,  0.14304489],\n",
            "       [ 0.06847109,  0.12520733,  0.21491683, ...,  0.00679054,\n",
            "        -0.1281595 ,  0.19423892],\n",
            "       [ 0.20965493,  0.08622533,  0.16823763, ..., -0.13317753,\n",
            "        -0.10735869,  0.10826949]], dtype=float32), 'protein_embedding': array([ 0.05058639,  0.06812885,  0.09517762, ...,  0.03887784,\n",
            "       -0.04277791,  0.11117269], dtype=float32)}, 'MACHINELEARNINA': {'residue_embeddings': array([[ 0.05924753, -0.03087843,  0.19369511, ...,  0.24348398,\n",
            "        -0.01742326, -0.01580246],\n",
            "       [ 0.00390356,  0.0963577 ,  0.06364392, ...,  0.02518233,\n",
            "        -0.07836936,  0.09168428],\n",
            "       [ 0.07678882,  0.04873591,  0.02363956, ..., -0.14371325,\n",
            "        -0.21987586, -0.03377383],\n",
            "       ...,\n",
            "       [-0.01647035,  0.1172629 ,  0.04350462, ..., -0.16278292,\n",
            "         0.13811204, -0.01609199],\n",
            "       [ 0.01158945,  0.0941322 ,  0.20851605, ..., -0.06242924,\n",
            "        -0.08559524,  0.04357257],\n",
            "       [ 0.06121174,  0.09304979, -0.04278693, ..., -0.15986326,\n",
            "        -0.07044572, -0.03227579]], dtype=float32), 'protein_embedding': array([ 0.01371042,  0.05903639,  0.09789606, ..., -0.04752282,\n",
            "       -0.01464671, -0.00314625], dtype=float32)}, 'MACHINELEARNINL': {'residue_embeddings': array([[ 0.00961785, -0.05928179,  0.05812462, ...,  0.27745146,\n",
            "        -0.13366118,  0.13794443],\n",
            "       [-0.00362006,  0.07556372,  0.041888  , ..., -0.00070747,\n",
            "        -0.04199704,  0.25303435],\n",
            "       [ 0.06700656,  0.07121369, -0.04138522, ..., -0.11637478,\n",
            "        -0.21843882,  0.08297256],\n",
            "       ...,\n",
            "       [-0.02776236,  0.01517877, -0.07946797, ..., -0.14941719,\n",
            "         0.10752523,  0.03953034],\n",
            "       [ 0.05766252,  0.06805132,  0.08338039, ..., -0.10037971,\n",
            "        -0.17228451,  0.04302412],\n",
            "       [ 0.12661219,  0.10969035,  0.10048413, ..., -0.03615028,\n",
            "        -0.12517497,  0.07052597]], dtype=float32), 'protein_embedding': array([ 0.01088777,  0.02653166,  0.02902342, ..., -0.03601252,\n",
            "       -0.06138286,  0.08772174], dtype=float32)}}\n",
            "{'WT': 'MACHINELEARNING', 'M1_1': 'MGCHINELEARNING', 'M1_2': 'MLCHINELEARNING', 'M2_1': 'MACHINEAEARNING', 'M2_2': 'MACHINEGEARNING', 'M3_1': 'MACHINELEARNINA', 'M3_2': 'MACHINELEARNINL'}\n"
          ]
        }
      ],
      "source": [
        "#Exercise 0\n",
        "import pickle\n",
        "from Bio import SeqIO\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "seq_embed={}\n",
        "toy_seq={}\n",
        "with open('toy_esm2_embeddings.pkl', 'rb') as f:\n",
        "  data=pickle.load(f)\n",
        "  for seq_record in SeqIO.parse(\"toy_sequences.fasta\", \"fasta\"):\n",
        "      toy_seq[str(seq_record.id)]=str(seq_record.seq)\n",
        "      seq_embed[str(seq_record.seq)]=data[seq_record.id]\n",
        "print(seq_embed)\n",
        "print(toy_seq)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Exercise 1\n",
        "residue_shape=seq_embed[toy_seq['WT']]['residue_embeddings'].shape\n",
        "print(f'Shape of residue ESM embedding: {residue_shape}')\n",
        "protein_shape=seq_embed[toy_seq['WT']]['protein_embedding'].shape\n",
        "print(f'Shape of protein ESM embedding: {protein_shape}')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "38y4blYTRq0i",
        "outputId": "4b20ff32-82e0-4de2-8331-99c7d8a7dc74"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Shape of residue ESM embedding: (15, 1280)\n",
            "Shape of protein ESM embedding: (1280,)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Exercise 1\n",
        "\n",
        "Q2: The shape is is directly proportional to the length of the input protein sequence."
      ],
      "metadata": {
        "id": "soGdhQmQYPYB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Exercise 2\n",
        "WT_mean=np.mean(seq_embed[toy_seq['WT']]['residue_embeddings'], axis=0)\n",
        "print(WT_mean)\n",
        "print(seq_embed[toy_seq['WT']]['protein_embedding'])\n",
        "\n",
        "assert list(WT_mean) == list(seq_embed[toy_seq['WT']]['protein_embedding'])\n",
        "print('They are the same')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "j9KGt_CHYXf-",
        "outputId": "e648fce7-70ab-4459-9237-e5cefc25154e"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[ 2.7157145e-02  6.7918792e-02  1.1835212e-01 ... -3.1786315e-02\n",
            "  6.4118205e-05  1.0375285e-02]\n",
            "[ 2.7157145e-02  6.7918792e-02  1.1835212e-01 ... -3.1786315e-02\n",
            "  6.4118205e-05  1.0375285e-02]\n",
            "They are the same\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Exercise 2\n",
        "\n",
        "Q1: An embedding per protein would lose structural and functional data that each amino acid contains since the protein embedding is an average into a vector.\n",
        "\n",
        "Q2:\n",
        "A single embedding for an entire protein allows for an efficient summary of the protein. It makes large scale analysis and machine learning more efficient."
      ],
      "metadata": {
        "id": "ZzCtbZpca2d-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Exercise 3\n",
        "WT_entry={}\n",
        "MT_entry={}\n",
        "with open('toy_esm2_embeddings.pkl', 'rb') as f:\n",
        "  data=pickle.load(f)\n",
        "  for key, value in data.items():\n",
        "    print(f'ID: {key}')\n",
        "    if key=='WT':\n",
        "      WT_entry[key]=value\n",
        "    else:\n",
        "      MT_entry[key]=value\n",
        "\n",
        "MT_sequences={}\n",
        "WT_sequences={}\n",
        "with open(\"toy_sequences.fasta\", \"r\") as f:\n",
        "    for record in SeqIO.parse(f, \"fasta\"):\n",
        "      if record.id !='WT':\n",
        "        MT_sequences[record.id]=str(record.seq)\n",
        "      else:\n",
        "        WT_sequences[record.id]=str(record.seq)\n",
        "\n",
        "for key,value in MT_sequences.items():\n",
        "  difference=0\n",
        "  for i in range(len(value)):\n",
        "    if value[i]!=WT_sequences['WT'][i]:\n",
        "      difference+=1\n",
        "    else:\n",
        "      continue\n",
        "  print(f'{key} and WT differ by {difference} chars')\n",
        "\n",
        "print(WT_entry)\n",
        "print(MT_entry)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pleKgwf5bmGp",
        "outputId": "ec8b888a-9e5f-472f-bfbb-b2064f21aec5"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ID: WT\n",
            "ID: M1_1\n",
            "ID: M1_2\n",
            "ID: M2_1\n",
            "ID: M2_2\n",
            "ID: M3_1\n",
            "ID: M3_2\n",
            "M1_1 and WT differ by 1 chars\n",
            "M1_2 and WT differ by 1 chars\n",
            "M2_1 and WT differ by 1 chars\n",
            "M2_2 and WT differ by 1 chars\n",
            "M3_1 and WT differ by 1 chars\n",
            "M3_2 and WT differ by 1 chars\n",
            "{'WT': {'residue_embeddings': array([[ 0.05606826, -0.02779197,  0.1984435 , ...,  0.29308623,\n",
            "        -0.0005211 , -0.001761  ],\n",
            "       [ 0.01368738,  0.08926699,  0.07153449, ...,  0.02676891,\n",
            "        -0.05958881,  0.08751322],\n",
            "       [ 0.0791374 ,  0.05739338,  0.05229965, ..., -0.12164824,\n",
            "        -0.2122045 , -0.03057704],\n",
            "       ...,\n",
            "       [-0.07602476,  0.17316677,  0.04309352, ..., -0.17556725,\n",
            "         0.18471679,  0.06163794],\n",
            "       [ 0.0032055 ,  0.07336932,  0.23836282, ..., -0.11572927,\n",
            "        -0.1422069 ,  0.17654972],\n",
            "       [ 0.1213042 ,  0.09664907,  0.15444879, ..., -0.22865772,\n",
            "        -0.07566334, -0.07008101]], dtype=float32), 'protein_embedding': array([ 2.7157145e-02,  6.7918792e-02,  1.1835212e-01, ...,\n",
            "       -3.1786315e-02,  6.4118205e-05,  1.0375285e-02], dtype=float32)}}\n",
            "{'M1_1': {'residue_embeddings': array([[-0.01221433, -0.04243858,  0.23617591, ...,  0.26975685,\n",
            "        -0.10064612,  0.03309401],\n",
            "       [ 0.1237796 , -0.01890814,  0.12185148, ...,  0.13858302,\n",
            "        -0.02139318,  0.07952821],\n",
            "       [ 0.0703975 , -0.04017996,  0.037202  , ..., -0.04747648,\n",
            "        -0.27260008,  0.03603963],\n",
            "       ...,\n",
            "       [-0.03181905,  0.18723276,  0.03050748, ..., -0.13183856,\n",
            "         0.15668514,  0.08637113],\n",
            "       [ 0.04160677,  0.09825741,  0.22365591, ..., -0.04876086,\n",
            "        -0.18204392,  0.16523133],\n",
            "       [ 0.11441068,  0.05866617,  0.15189052, ..., -0.16161145,\n",
            "        -0.11273676,  0.0340802 ]], dtype=float32), 'protein_embedding': array([ 0.03928996,  0.05962785,  0.12646566, ...,  0.01712848,\n",
            "       -0.03295295,  0.06646581], dtype=float32)}, 'M1_2': {'residue_embeddings': array([[-0.02348768, -0.06737775,  0.30448887, ...,  0.22063273,\n",
            "        -0.0216808 , -0.03832598],\n",
            "       [ 0.1255877 , -0.02136015,  0.20788512, ...,  0.07072802,\n",
            "         0.02024508, -0.0259651 ],\n",
            "       [ 0.13519073, -0.08404002,  0.0928096 , ..., -0.09306771,\n",
            "        -0.29415426,  0.03936787],\n",
            "       ...,\n",
            "       [-0.04450252,  0.18036602,  0.06166712, ..., -0.17303912,\n",
            "         0.16002263,  0.09177554],\n",
            "       [ 0.02496321,  0.1156003 ,  0.24694373, ..., -0.0935742 ,\n",
            "        -0.16940343,  0.20230731],\n",
            "       [ 0.08165168,  0.09079417,  0.12927775, ..., -0.1996663 ,\n",
            "        -0.06689113,  0.01048542]], dtype=float32), 'protein_embedding': array([ 0.04051566,  0.05903596,  0.15929994, ..., -0.01560776,\n",
            "       -0.02459794,  0.06557129], dtype=float32)}, 'M2_1': {'residue_embeddings': array([[ 0.04304859, -0.01717667,  0.20790304, ...,  0.28659803,\n",
            "        -0.02484175,  0.02266391],\n",
            "       [-0.01545749,  0.10129584,  0.05071112, ...,  0.03306917,\n",
            "        -0.08175289,  0.11969808],\n",
            "       [ 0.09800516,  0.04179817,  0.0534738 , ..., -0.13397202,\n",
            "        -0.22516078,  0.013848  ],\n",
            "       ...,\n",
            "       [-0.05305982,  0.2048952 ,  0.05135461, ..., -0.17196594,\n",
            "         0.19058347,  0.09969407],\n",
            "       [ 0.0415617 ,  0.12278973,  0.2219481 , ..., -0.09047426,\n",
            "        -0.15482979,  0.18269931],\n",
            "       [ 0.11829355,  0.08116539,  0.1686621 , ..., -0.22043261,\n",
            "        -0.10947494,  0.01592181]], dtype=float32), 'protein_embedding': array([ 0.03536478,  0.08991922,  0.11587524, ..., -0.02221408,\n",
            "       -0.01809251,  0.06115762], dtype=float32)}, 'M2_2': {'residue_embeddings': array([[-0.02047198, -0.05934278,  0.18795408, ...,  0.31268495,\n",
            "        -0.07668409,  0.04833537],\n",
            "       [-0.05122827,  0.09534466, -0.01075379, ...,  0.06221346,\n",
            "        -0.1297039 ,  0.14222705],\n",
            "       [ 0.0700347 ,  0.01101001,  0.0208775 , ..., -0.1051219 ,\n",
            "        -0.25953165,  0.08852682],\n",
            "       ...,\n",
            "       [-0.00587133,  0.173598  ,  0.04168514, ..., -0.07634631,\n",
            "         0.17283033,  0.14304489],\n",
            "       [ 0.06847109,  0.12520733,  0.21491683, ...,  0.00679054,\n",
            "        -0.1281595 ,  0.19423892],\n",
            "       [ 0.20965493,  0.08622533,  0.16823763, ..., -0.13317753,\n",
            "        -0.10735869,  0.10826949]], dtype=float32), 'protein_embedding': array([ 0.05058639,  0.06812885,  0.09517762, ...,  0.03887784,\n",
            "       -0.04277791,  0.11117269], dtype=float32)}, 'M3_1': {'residue_embeddings': array([[ 0.05924753, -0.03087843,  0.19369511, ...,  0.24348398,\n",
            "        -0.01742326, -0.01580246],\n",
            "       [ 0.00390356,  0.0963577 ,  0.06364392, ...,  0.02518233,\n",
            "        -0.07836936,  0.09168428],\n",
            "       [ 0.07678882,  0.04873591,  0.02363956, ..., -0.14371325,\n",
            "        -0.21987586, -0.03377383],\n",
            "       ...,\n",
            "       [-0.01647035,  0.1172629 ,  0.04350462, ..., -0.16278292,\n",
            "         0.13811204, -0.01609199],\n",
            "       [ 0.01158945,  0.0941322 ,  0.20851605, ..., -0.06242924,\n",
            "        -0.08559524,  0.04357257],\n",
            "       [ 0.06121174,  0.09304979, -0.04278693, ..., -0.15986326,\n",
            "        -0.07044572, -0.03227579]], dtype=float32), 'protein_embedding': array([ 0.01371042,  0.05903639,  0.09789606, ..., -0.04752282,\n",
            "       -0.01464671, -0.00314625], dtype=float32)}, 'M3_2': {'residue_embeddings': array([[ 0.00961785, -0.05928179,  0.05812462, ...,  0.27745146,\n",
            "        -0.13366118,  0.13794443],\n",
            "       [-0.00362006,  0.07556372,  0.041888  , ..., -0.00070747,\n",
            "        -0.04199704,  0.25303435],\n",
            "       [ 0.06700656,  0.07121369, -0.04138522, ..., -0.11637478,\n",
            "        -0.21843882,  0.08297256],\n",
            "       ...,\n",
            "       [-0.02776236,  0.01517877, -0.07946797, ..., -0.14941719,\n",
            "         0.10752523,  0.03953034],\n",
            "       [ 0.05766252,  0.06805132,  0.08338039, ..., -0.10037971,\n",
            "        -0.17228451,  0.04302412],\n",
            "       [ 0.12661219,  0.10969035,  0.10048413, ..., -0.03615028,\n",
            "        -0.12517497,  0.07052597]], dtype=float32), 'protein_embedding': array([ 0.01088777,  0.02653166,  0.02902342, ..., -0.03601252,\n",
            "       -0.06138286,  0.08772174], dtype=float32)}}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Exercise 4\n",
        "from sklearn.metrics.pairwise import cosine_similarity\n",
        "\n",
        "\n",
        "def compute_similarity(WT_embedding, MT_embedding):\n",
        "  queryid=['WT']\n",
        "  query_matrix=np.array([WT_embedding[id] for id in queryid])\n",
        "\n",
        "  db_ids=list(MT_embedding.keys())\n",
        "  db_matrix=np.array([MT_embedding[id] for id in db_ids])\n",
        "\n",
        "  comp=cosine_similarity(query_matrix,db_matrix)\n",
        "  return comp\n",
        "def tophits(similarity, MT_embedding,MT_info):\n",
        "  MT_ids=list(MT_embedding.keys())\n",
        "  result=[]\n",
        "\n",
        "  for i in range(len(similarity[0])):\n",
        "      hit_id=MT_ids[i]\n",
        "      score=similarity[0][i]\n",
        "      result.append(\n",
        "          {\n",
        "          'sequence id': hit_id,\n",
        "          'cosine similarity': score,\n",
        "          'Mutation position': MT_info[hit_id][0],\n",
        "          'Amino acid change': MT_info[hit_id][2]+'--->'+MT_info[hit_id][1]\n",
        "    }\n",
        "      )\n",
        "  return result\n",
        "\n",
        "MT_info={}\n",
        "MT_embedding={}\n",
        "WT_embedding={}\n",
        "for key,value in MT_sequences.items():\n",
        "  MT_embedding[key]=MT_entry[key]['protein_embedding']\n",
        "\n",
        "  for i in range(len(value)):\n",
        "    if value[i]!=WT_sequences['WT'][i]:\n",
        "\n",
        "      #list structured as [idx,mutant char,wt char]\n",
        "      MT_info[key]=[i,value[i],WT_sequences['WT'][i]]\n",
        "      print(f'{key} differs from WT at index pos {i}, where {WT_sequences['WT'][i]} is swapped with {value[i]}')\n",
        "    else:\n",
        "      continue\n",
        "\n",
        "WT_embedding['WT']=WT_entry['WT']['protein_embedding']\n",
        "\n",
        "comparison=compute_similarity(WT_embedding,MT_embedding)\n",
        "table=tophits(comparison,MT_embedding,MT_info)\n",
        "resultdf=pd.DataFrame(table)\n",
        "display(resultdf)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 359
        },
        "id": "J5M1YH7O7d06",
        "outputId": "bad62fb8-36a9-4d74-9284-0dd401472c40"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "M1_1 differs from WT at index pos 1, where A is swapped with G\n",
            "M1_2 differs from WT at index pos 1, where A is swapped with L\n",
            "M2_1 differs from WT at index pos 7, where L is swapped with A\n",
            "M2_2 differs from WT at index pos 7, where L is swapped with G\n",
            "M3_1 differs from WT at index pos 14, where G is swapped with A\n",
            "M3_2 differs from WT at index pos 14, where G is swapped with L\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "  sequence id  cosine similarity  Mutation position Amino acid change\n",
              "0        M1_1           0.993105                  1            A--->G\n",
              "1        M1_2           0.990491                  1            A--->L\n",
              "2        M2_1           0.997603                  7            L--->A\n",
              "3        M2_2           0.988023                  7            L--->G\n",
              "4        M3_1           0.997718                 14            G--->A\n",
              "5        M3_2           0.983698                 14            G--->L"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0fe4676e-505f-40b3-ad8e-1d42b753e39b\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sequence id</th>\n",
              "      <th>cosine similarity</th>\n",
              "      <th>Mutation position</th>\n",
              "      <th>Amino acid change</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>M1_1</td>\n",
              "      <td>0.993105</td>\n",
              "      <td>1</td>\n",
              "      <td>A---&gt;G</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>M1_2</td>\n",
              "      <td>0.990491</td>\n",
              "      <td>1</td>\n",
              "      <td>A---&gt;L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>M2_1</td>\n",
              "      <td>0.997603</td>\n",
              "      <td>7</td>\n",
              "      <td>L---&gt;A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>M2_2</td>\n",
              "      <td>0.988023</td>\n",
              "      <td>7</td>\n",
              "      <td>L---&gt;G</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>M3_1</td>\n",
              "      <td>0.997718</td>\n",
              "      <td>14</td>\n",
              "      <td>G---&gt;A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>M3_2</td>\n",
              "      <td>0.983698</td>\n",
              "      <td>14</td>\n",
              "      <td>G---&gt;L</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0fe4676e-505f-40b3-ad8e-1d42b753e39b')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0fe4676e-505f-40b3-ad8e-1d42b753e39b button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0fe4676e-505f-40b3-ad8e-1d42b753e39b');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "  <div id=\"id_cf0cd8b7-141c-4d46-b984-9ec5c1ed244e\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('resultdf')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_cf0cd8b7-141c-4d46-b984-9ec5c1ed244e button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('resultdf');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "resultdf",
              "summary": "{\n  \"name\": \"resultdf\",\n  \"rows\": 6,\n  \"fields\": [\n    {\n      \"column\": \"sequence id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"M1_1\",\n          \"M1_2\",\n          \"M3_2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cosine similarity\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          0.993104875087738,\n          0.9904909133911133,\n          0.983698308467865\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Mutation position\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5,\n        \"min\": 1,\n        \"max\": 14,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          1,\n          7,\n          14\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Amino acid change\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"A--->G\",\n          \"A--->L\",\n          \"G--->L\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    }
  ]
}